src/twisted/conch/ssh/session.py

```diff
@@ -180,10 +180,14 @@ def extReceived(self, dataType, data):
             log.warn("Weird extended data: {dataType}", dataType=dataType)
 
     def eofReceived(self):
-        if self.client:
-            self.conn.sendClose(self)
+        # If we have a session, tell it that EOF has been received and
```

Comment:
Yeah, your approach broadly seems better - it's just the difficult compatibility issues of getting there from here.  As I mentioned in my previous PR (#1696), the most practical approach may be to move the EOF/close handling into the process protocol (`self.client`) and just create an unconnected process protocol earlier for cases such as environment variable passing requests in advance of setting up the real session.  I don't think this PR would make that plan any harder to execute; it would just need the corresponding change to the client's `eofReceived` to only call `conn.sendClose` if it doesn't have a session.

But for now, this restores previous behaviour ...